modules/healthcheck/default.nix

```
@@ -0,0 +1,281 @@
+                            while true; do
+                              local output
+                              output=$(sh -c "${probeCfg.command}" 2>/dev/null)
+
+                              if [[ -n "''${output}" ]]; then
+                                echo "Health check: probe successful. Notifying systemd that service is ready."
+                                systemd-notify --ready --status="${successStatus}"
+                                return 0
+                              else
+                                echo "Health check: probe not successful. Notifying systemd that service is still waiting."
+                                systemd-notify --status="${probeCfg.statusWaiting}"
+                              fi
+                              sleep ${toString probeCfg.periodSeconds}
+                            done
+                          }
+
+                          if [[ -n "''${NOTIFY_SOCKET:-}" ]]; then
+                            check &
+                          else
+                            echo "Health check: NOTIFY_SOCKET not set. Cannot run readiness probe." >&2
+                          fi
+
+                          ${cfg.exec}
+                        ''
+                      );
+                  }
+                ))
+              ) servicesWithHealthcheck)
+              // lib.mapAttrs' (
+                mainServiceName: serviceConfig:
+                let
+                  cfg = serviceConfig.healthcheck;
+                in
+                {
+                  name = "${mainServiceName}-liveness-check";
+                  value = mkIf (cfg != null && cfg.liveness-probe.enable) (
+                    let
+                      probe = cfg.liveness-probe;
+                      checkScript = pkgs.writeShellScriptBin "liveness-check" ''
+                        #!${pkgs.runtimeShell}
+                        sleep ${toString probe.initialDelaySeconds}
+                        echo "Executing liveness probe for ${mainServiceName}..."
+                        # If the command fails, explicitly restart the main service
+                        while true; do
+                          if ! (timeout ${toString probe.timeoutSeconds}s ${probe.command}); then
+                            echo "(timeout ${toString probe.timeoutSeconds}s ${probe.command})"
+                            echo "Liveness probe for ${mainServiceName} failed. Triggering restart..."
+                            ${pkgs.systemd}/bin/systemctl restart ${mainServiceName}.service &
+                          fi
+                          sleep ${toString probe.periodSeconds}
+                        done
+                      '';
+                    in
+                    {
+                      description = "Liveness check for ${mainServiceName}";
+                      # This check needs systemctl in its path.
+                      path = [ pkgs.systemd ] ++ (if cfg ? "path" then cfg.path else [ ]);
```

Maybe prefer [`lib.optionals`](https://noogle.dev/f/lib/optionals)

```suggestion
                      path = [ pkgs.systemd ] ++ (lib.optionals (cfg ? "path") cfg.path);
```

or maybe this, as per <https://nix.dev/manual/nix/2.26/language/operators#attribute-selection>

```suggestion
                      path = [ pkgs.systemd ] ++ (cfg.path or [ ]);
```